In [2]:
import numpy as np
import pandas as pd

In [3]:
df_train = pd.read_csv('train.csv')
df_train.head()

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     1     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 378 columns]

In [4]:
df_train.shape

(4209, 378)

In [5]:
df_train.describe()

ID            y          X10     X11          X12  \
count  4209.000000  4209.000000  4209.000000  4209.0  4209.000000   
mean   4205.960798   100.669318     0.013305     0.0     0.075077   
std    2437.608688    12.679381     0.114590     0.0     0.263547   
min       0.000000    72.110000     0.000000     0.0     0.000000   
25%    2095.000000    90.820000     0.000000     0.0     0.000000   
50%    4220.000000    99.150000     0.000000     0.0     0.000000   
75%    6314.000000   109.010000     0.000000     0.0     0.000000   
max    8417.000000   265.320000     1.000000     0.0     1.000000   

               X13          X14          X15          X16          X17  ...  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  ...   
mean      0.057971     0.428130     0.000475     0.002613     0.007603  ...   
std       0.233716     0.494867     0.021796     0.051061     0.086872  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       0.000000     1.000000     0.000000     0.000000     0.000000  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

              X375         X376         X377         X378         X379  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000   
mean      0.318841     0.057258     0.314802     0.020670     0.009503   
std       0.466082     0.232363     0.464492     0.142294     0.097033   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

              X380         X382         X383         X384         X385  
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  
mean      0.008078     0.007603     0.001663     0.000475     0.001426  
std       0.089524     0.086872     0.040752     0.021796     0.037734  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 370 columns]

In [6]:
features = df_train.drop(['ID', 'y'], axis=1)

print(features.shape)

features.dtypes.value_counts()

(4209, 376)


int64     368
object      8
dtype: int64

In [7]:
df_test = pd.read_csv('test.csv')

In [8]:
df_test.shape

(4209, 377)

In [9]:
df_test.head()

ID  X0 X1  X2 X3 X4 X5 X6 X8  X10  ...  X375  X376  X377  X378  X379  X380  \
0   1  az  v   n  f  d  t  a  w    0  ...     0     0     0     1     0     0   
1   2   t  b  ai  a  d  b  g  y    0  ...     0     0     1     0     0     0   
2   3  az  v  as  f  d  a  j  j    0  ...     0     0     0     1     0     0   
3   4  az  l   n  f  d  z  l  n    0  ...     0     0     0     1     0     0   
4   5   w  s  as  c  d  y  i  m    0  ...     1     0     0     0     0     0   

   X382  X383  X384  X385  
0     0     0     0     0  
1     0     0     0     0  
2     0     0     0     0  
3     0     0     0     0  
4     0     0     0     0  

[5 rows x 377 columns]

In [10]:
df_test.describe()

ID          X10          X11          X12          X13  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000   
mean   4211.039202     0.019007     0.000238     0.074364     0.061060   
std    2423.078926     0.136565     0.015414     0.262394     0.239468   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%    2115.000000     0.000000     0.000000     0.000000     0.000000   
50%    4202.000000     0.000000     0.000000     0.000000     0.000000   
75%    6310.000000     0.000000     0.000000     0.000000     0.000000   
max    8416.000000     1.000000     1.000000     1.000000     1.000000   

               X14          X15          X16          X17          X18  ...  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  ...   
mean      0.427893     0.000713     0.002613     0.008791     0.010216  ...   
std       0.494832     0.026691     0.051061     0.093357     0.100570  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       1.000000     0.000000     0.000000     0.000000     0.000000  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

              X375         X376         X377         X378         X379  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000   
mean      0.325968     0.049656     0.311951     0.019244     0.011879   
std       0.468791     0.217258     0.463345     0.137399     0.108356   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

              X380         X382         X383         X384         X385  
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  
mean      0.008078     0.008791     0.000475     0.000713     0.001663  
std       0.089524     0.093357     0.021796     0.026691     0.040752  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 369 columns]

In [11]:
X_train = features
X_test = df_test[features.columns]

y_train = df_train['y']

id_test = df_test['ID']

In [12]:
#Variance = 0

for col in X_train.columns:
    if X_train[col].nunique() == 1:
        del X_train[col]
        del X_test[col]


Null Values

In [13]:
X_train.isnull().any().value_counts()

False    364
dtype: int64

In [14]:
X_test.isnull().any().value_counts()

False    364
dtype: int64

Unique Values

In [15]:
#X_train["X21"].nunique()

unique_train = X_train.apply(lambda x: x.nunique())
unique_train

X0      47
X1      27
X2      44
X3       7
X4       4
        ..
X380     2
X382     2
X383     2
X384     2
X385     2
Length: 364, dtype: int64

In [16]:
unique_test = X_test.apply(lambda x: x.nunique())
unique_test

X0      49
X1      27
X2      45
X3       7
X4       4
        ..
X380     2
X382     2
X383     2
X384     2
X385     2
Length: 364, dtype: int64

In [17]:
X_train.shape, X_test.shape

((4209, 364), (4209, 364))

In [18]:
#Experimental

We concatenate both dataframes because the categorical features in both the sets contain some differrent values and label encoding them seperately may cause unseen issues.

In [19]:
X_both = pd.concat([X_train, X_test])
X_both

X0  X1  X2 X3 X4  X5 X6 X8  X10  X12  ...  X375  X376  X377  X378  X379  \
0      k   v  at  a  d   u  j  o    0    0  ...     0     0     1     0     0   
1      k   t  av  e  d   y  l  o    0    0  ...     1     0     0     0     0   
2     az   w   n  c  d   x  j  x    0    0  ...     0     0     0     0     0   
3     az   t   n  f  d   x  l  e    0    0  ...     0     0     0     0     0   
4     az   v   n  f  d   h  d  n    0    0  ...     0     0     0     0     0   
...   ..  ..  .. .. ..  .. .. ..  ...  ...  ...   ...   ...   ...   ...   ...   
4204  aj   h  as  f  d  aa  j  e    0    0  ...     0     0     0     0     0   
4205   t  aa  ai  d  d  aa  j  y    0    0  ...     0     1     0     0     0   
4206   y   v  as  f  d  aa  d  w    0    0  ...     0     0     0     0     0   
4207  ak   v  as  a  d  aa  c  q    0    0  ...     0     0     1     0     0   
4208   t  aa  ai  c  d  aa  g  r    0    0  ...     1     0     0     0     0   

      X380  X382  X383  X384  X385  
0        0     0     0     0     0  
1        0     0     0     0     0  
2        0     1     0     0     0  
3        0     0     0     0     0  
4        0     0     0     0     0  
...    ...   ...   ...   ...   ...  
4204     0     0     0     0     0  
4205     0     0     0     0     0  
4206     0     0     0     0     0  
4207     0     0     0     0     0  
4208     0     0     0     0     0  

[8418 rows x 364 columns]

In [20]:
#Label encoder
from sklearn import preprocessing 
le = preprocessing.LabelEncoder()

In [21]:
for col in X_both.select_dtypes(include=['object']):
    X_both[col] = le.fit_transform(X_both[col])
X_both.head()

X0  X1  X2  X3  X4  X5  X6  X8  X10  X12  ...  X375  X376  X377  X378  \
0  37  23  20   0   3  27   9  14    0    0  ...     0     0     1     0   
1  37  21  22   4   3  31  11  14    0    0  ...     1     0     0     0   
2  24  24  38   2   3  30   9  23    0    0  ...     0     0     0     0   
3  24  21  38   5   3  30  11   4    0    0  ...     0     0     0     0   
4  24  23  38   5   3  14   3  13    0    0  ...     0     0     0     0   

   X379  X380  X382  X383  X384  X385  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     1     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 364 columns]

In [22]:
X_train_le = X_both[:4209]
X_train_le.head()

X0  X1  X2  X3  X4  X5  X6  X8  X10  X12  ...  X375  X376  X377  X378  \
0  37  23  20   0   3  27   9  14    0    0  ...     0     0     1     0   
1  37  21  22   4   3  31  11  14    0    0  ...     1     0     0     0   
2  24  24  38   2   3  30   9  23    0    0  ...     0     0     0     0   
3  24  21  38   5   3  30  11   4    0    0  ...     0     0     0     0   
4  24  23  38   5   3  14   3  13    0    0  ...     0     0     0     0   

   X379  X380  X382  X383  X384  X385  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     1     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 364 columns]

In [23]:
X_test_le = X_both[4209:]
X_test_le.head()

X0  X1  X2  X3  X4  X5  X6  X8  X10  X12  ...  X375  X376  X377  X378  \
0  24  23  38   5   3  26   0  22    0    0  ...     0     0     0     1   
1  46   3   9   0   3   9   6  24    0    0  ...     0     0     1     0   
2  24  23  19   5   3   0   9   9    0    0  ...     0     0     0     1   
3  24  13  38   5   3  32  11  13    0    0  ...     0     0     0     1   
4  49  20  19   2   3  31   8  12    0    0  ...     1     0     0     0   

   X379  X380  X382  X383  X384  X385  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     0     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 364 columns]

In [24]:
from sklearn.decomposition import PCA

In [25]:
# We retain more than 99 percent of the Explained variance
pca = PCA(0.99, random_state=42)
Pca_train = pca.fit_transform(X_train_le)
Pca_test = pca.transform(X_test_le)

In [26]:
Pca_train.shape , Pca_test.shape

((4209, 24), (4209, 24))

We reduced the number of features from 364 to 24 while retaining more than 99% of explained variance

In [27]:
import xgboost as xgb

D:\Anaconda_2022_installation\lib\site-packages\xgboost\compat.py:36: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import MultiIndex, Int64Index


In [28]:
from sklearn.model_selection import GridSearchCV

In [33]:
xgb_model = xgb.XGBRegressor()

In [47]:
params = {'learning_rate' : [0.025,0.05,0.1],
         'max_depth' : [2,3,4],
          'gamma' : [0,0.01,0.02]
         }

In [48]:
clf = GridSearchCV(xgb_model, params, scoring ='r2', verbose = 2 , n_jobs = -1)
clf.fit(X_train_le, y_train)

Fitting 5 folds for each of 27 candidates, totalling 135 fits


GridSearchCV(estimator=XGBRegressor(base_score=0.5, booster='gbtree',
                                    colsample_bylevel=1, colsample_bynode=1,
                                    colsample_bytree=1,
                                    enable_categorical=False, gamma=0,
                                    gpu_id=-1, importance_type=None,
                                    interaction_constraints='',
                                    learning_rate=0.300000012, max_delta_step=0,
                                    max_depth=6, min_child_weight=1,
                                    missing=nan, monotone_constraints='()',
                                    n_estimators=100, n_jobs=4,
                                    num_parallel_tree=1, predictor='auto',
                                    random_state=0, reg_alpha=0, reg_lambda=1,
                                    scale_pos_weight=1, subsample=1,
                                    tree_method='exact', validate_parameters=1,


In [49]:
clf.best_params_

{'gamma': 0, 'learning_rate': 0.05, 'max_depth': 3}

In [50]:
clf.best_score_

0.5680572005495828

In [52]:
xgb_model = xgb.XGBRegressor(params = {'gamma': 0, 'learning_rate': 0.05, 'max_depth': 3})

In [53]:
xgb_model.fit(X_train_le, y_train)

D:\Anaconda_2022_installation\lib\site-packages\xgboost\data.py:262: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  elif isinstance(data.columns, (pd.Int64Index, pd.RangeIndex)):


[17:51:05] WARNING: D:\bld\xgboost-split_1645118015404\work\src\learner.cc:576: 
Parameters: { "params" } might not be used.

  This could be a false alarm, with some parameters getting used by language bindings but
  then being mistakenly passed down to XGBoost core, or some parameter actually being used
  but getting flagged wrongly here. Please open an issue if you find any such cases.




XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, enable_categorical=False,
             gamma=0, gpu_id=-1, importance_type=None,
             interaction_constraints='', learning_rate=0.300000012,
             max_delta_step=0, max_depth=6, min_child_weight=1, missing=nan,
             monotone_constraints='()', n_estimators=100, n_jobs=4,
             num_parallel_tree=1,
             params={'gamma': 0, 'learning_rate': 0.05, 'max_depth': 3},
             predictor='auto', random_state=0, reg_alpha=0, reg_lambda=1,
             scale_pos_weight=1, subsample=1, tree_method='exact',
             validate_parameters=1, verbosity=None)

In [54]:
xgb_model.score(X_train_le,y_train)

D:\Anaconda_2022_installation\lib\site-packages\xgboost\data.py:262: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  elif isinstance(data.columns, (pd.Int64Index, pd.RangeIndex)):


0.8487697345100329

In [56]:
y_pred = xgb_model.predict(X_test_le)

D:\Anaconda_2022_installation\lib\site-packages\xgboost\data.py:262: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  elif isinstance(data.columns, (pd.Int64Index, pd.RangeIndex)):


In [57]:
MB = pd.DataFrame()
MB['Id'] = id_test
MB['y'] = y_pred
MB.to_csv('mb.csv', index=False)

In [58]:
MB.head()

Id           y
0   1  100.342232
1   2  115.859055
2   3  100.843636
3   4   80.435280
4   5  111.313156